Apply keys transformation only on plain objects

ben-eb opened the pull request:
Closes https://github.com/blakeembrey/change-case/issues/356

ben-eb committed:
Apply keys transformation only on plain objects

TrySound commented on packages/change-case/package.json:
```diff
@@ -45,6 +45,9 @@
     "test": "ts-scripts test"
   },
+  "dependencies": {
+    "is-plain-obj": "^4.1.0"
```
I would inline it to not introduce extra dependencies.

Btw hi dude!

ben-eb replied:
Updated 🙂 

Hope you're doing well @TrySound !

ben-eb committed:
Inline isPlainObject check

Revadike commented:
@TrySound can this be merged?

Btw, I'm curious, from what I can tell this project has been around for nearly 12 years, how is it only now this problem has surfaced?

blakeembrey commented:
> how is it only now this problem has surfaced?

It was added in 2023 with https://github.com/blakeembrey/change-case/issues/86#issuecomment-1741631382.

@ben-eb I'm a little busy with a newborn but I'll try to get around to releasing in the next month, thanks for the contribution!

blakeembrey commented on packages/change-case/src/keys.ts:
```diff
@@ -13,12 +25,14 @@ function changeKeysFactory<
       return object.map((item) => changeKeys(item, depth - 1, options));
     }
 
+    if (!isPlainObject(object)) return object;
```
Nit: We access `Object.getPrototypeOf` below too, wdyt about accessing it once and making a second function take the prototype to check it? E.g. `isObject(x) && isPlainPrototype(x)`?

ben-eb replied:
I could re-use the `Object.getPrototypeOf(object)` call and pass it into `isPlainObject`;

```typescript
function isPlainObject(prototype: unknown, value: unknown) {
  if (typeof value !== "object" || value === null) {
    return false;
  }

  return (
    (prototype === null ||
      prototype === Object.prototype ||
      Object.getPrototypeOf(prototype) === null) &&
    !(Symbol.toStringTag in value) &&
    !(Symbol.iterator in value)
  );
}

// ...

const prototype = Object.getPrototypeOf(object);

if (!isPlainObject(prototype, object)) return false;

const result = Object.create(prototype);
```

Not sure about this implementation because it requires the caller to know more about the internals of `isPlainObject`, rather than a standalone implementation. It may be slightly more performant though 🙂

ben-eb replied:
@blakeembrey Is this holding up landing this feature?

blakeembrey commented on packages/change-case/src/keys.spec.ts:
```diff
@@ -63,6 +63,7 @@ const TEST_CASES: TestCase[] = [
     },
     { mergeAmbiguousCharacters: true },
   ],
+  [new Date(), 1, new Date()],
```
Can we add tests for every condition we're checking in `isPlainObject`?

ben-eb replied:
Sure thing. Will try to find some time this week 🙂

ben-eb replied:
Added some more tests. 👍

ben-eb committed:
Prefer describe.each

This is so that we can test `Object.create(null)`. Currently,
that fails with `TypeError: Cannot convert object to primitive value`

ben-eb committed:
Create helper for testing identical input/output

ben-eb committed:
Add more tests for keys

ben-eb committed:
Fix variable in template string

wasied commented:
Hey, I have the same issue on my end. Any news on this? Thanks :)

pantajoe commented:
Hello there! I don't mean to press, but this seems like a promising fix and as I'm currently having the same issue, I'd love to see this fixed in an upcoming release 🫣
